KS-16-05-RPG

DenisGubernatorov opened the pull request:
1) Создан базовый класс для персонажа. В конструкторе можно установить базовые характеристики здоровья и силы удара и количества здоровья которое восстанавливается по умолчанию 2) Интерфейс BaseCharacterActions реализует базовые действия для всех персонажей. 3) Класс Warrior наследник Character и реализующий BaseCharacterActions

DenisGubernatorov committed:
KS-16-05-RPG
1) Создан базовый класс для персонажа. В конструкторе можно установить базовые характеристики здоровья и силы удара и количества здоровья которое восстанавливается по умолчанию
2) Интерфейс BaseCharacterActions реализует базовые действия для всех персонажей.
3) Класс Warrior наследник Character и реализующий BaseCharacterActions

DenisGubernatorov commented:
Main  реализован не очень красиво. Там можно было сделать через ввод какое действие совершить и для дамага дополнитльно вводить количество полученного урона. Но решил не усложнять в рамках задачи

ievetrovdev commented on src/main/kotlin/lesson16/lesson16_task5.kt:
```diff
@@ -1 +1,72 @@
+    override fun getDamage(getDamage: Int): Boolean {
+        println("$name  получает  $getDamage ед. урона")
+        health -= getDamage
+        return isDead(health)
```
Персонажа заставляют умереть при каждом получении урона. Надо бы вынести вызов этой функции в отдельное условие

ievetrovdev commented on src/main/kotlin/lesson16/lesson16_task5.kt:
```diff
@@ -1 +1,72 @@
+    protected var damage: Int,
+    protected var healVal: Int,
+) {
+    protected abstract fun isDead(param: Int): Boolean
```
По заданию: "Добавить **приватный** метод, который вызывается, если при получении урона заканчивается здоровье.", а тут protected

ievetrovdev commented on src/main/kotlin/lesson16/lesson16_task5.kt:
```diff
@@ -1 +1,72 @@
+     */
+    override fun getDamage(getDamage: Int): Boolean {
+        println("$name  получает  $getDamage ед. урона")
+        health -= getDamage
```
Следует вычитать из здоровья абсолютное значение дамага (есть функция).

DenisGubernatorov committed:
KS-16-05-RPG
Рефакторинг
1) Метод isDead сделан приватным с дефолтной реализацией в классе-родителе
2) Добавлен отдельный метод checkIsDead, для отдельной проверки умер персонаж или нет. Метод сделан доступным для переопределения в наследниках(вариант - класс Некромант, который после смерти может продолжить сражаться)
3) При получении урона используется абсолютное значение переданного урона

ievetrovdev commented on src/main/kotlin/lesson16/lesson16_task5.kt:
```diff
@@ -1 +1,73 @@
+    /**
+     * Количество полученного дамага от другого персонажа
+     */
+    override fun getDamage(getDamage: Int): Int {
```
Согласно ТЗ: "Добавить приватный метод, **который вызывается, если при получении урона заканчивается здоровье**." - Нужно добавить в этот метод checkIsDead. Получение урона должно автоматически убивать.

DenisGubernatorov replied:
Если честно, то не совсем понимаю. В первоначальном варинате и вызвался метод isDead из метода getDamage (https://github.com/DenisGubernatorov/KotlinSprint/pull/90#discussion_r1740201263). Единственный момент, что он был prtected, а не privat. Но ведь, по сути, protected мотоды видны в самом классе и в классах наследниках. Но вот обратиться к этому методу из экземпляра класса наследники нельзя. Что ответчает требованию приватности. А если не делать protected в классе родителе, то его нельзя переопределить в наследниках. И пришлось добавлять лишнюю, на мой взгляд, логику вызова через public метод.

ievetrovdev replied:
В целом, тут не обязательно было делать такую иерархию, да и само задание делает такое не очень удобным, впрочем, главное добиться того, что бы персонаж мог умереть при получении урона, без дополнительных манипуляций и методов

DenisGubernatorov committed:
KS-16-05-RPG
1) Вызов метода checkIsDead перенесен в метод getDamage

ievetrovdev reviewed (approved):
Все верно, отличное решение!
